### ISSUE TITLE
Non-mutually-exclusive Segmentation

### ISSUE THREAD

**Issue by zachcoleman** [open-mmlab/mmsegmentation#890]
# Describe the feature

**Motivation**
In a variety of use cases, it's practical to assign more than one class to a pixel and/or have class-wise logits returned so thresholding can be applied per-class. An example of this is classifying pixels as both an occluding and occluded class.  A simple example is trees that overhang or occlude roofs in which a model can learn and infer that the pixel is both a roof (occluded) and is a tree (occluding) rather than be restricted to classifying the pixel to a single class. 

Currently, `mmseg` is very close to supporting this schema if a careful combination of settings are used:

- Using `CrossEntropyLoss` with `sigmoid = True` (see: [here](https://github.com/open-mmlab/mmsegmentation/blob/d35fbbdb47383906399df3a23e4f9f1db1e68b2b/mmseg/models/losses/cross_entropy_loss.py#L169-L170))
- Modifying the input pipeline such that `gt_semantic_seg` is a 3D tensor matching the `seg_logits`. This works with torch's `binary_cross_entropy_with_logits` called [here](https://github.com/open-mmlab/mmsegmentation/blob/d35fbbdb47383906399df3a23e4f9f1db1e68b2b/mmseg/models/losses/cross_entropy_loss.py#L92-L93) by `binary_cross_entropy`. _Note: in some cases this 3D by 3D cross entropy calculation is already being done with one-hot expansion already_.
- Inferencing with `whole_inference` or `slide_inference` to avoid both  argmax-ing and sofmax-ing [here](https://github.com/open-mmlab/mmsegmentation/blob/d35fbbdb47383906399df3a23e4f9f1db1e68b2b/mmseg/models/segmentors/encoder_decoder.py#L257) and [here](https://github.com/open-mmlab/mmsegmentation/blob/d35fbbdb47383906399df3a23e4f9f1db1e68b2b/mmseg/models/segmentors/encoder_decoder.py#L242) respectively.
- Overriding the `DefaultFormatBundle`'s default behavior to unsqueeze the `gt_semantic_seg` so that 3D tensors are packaged correctly.

There is one blocker however:
- When the losses are collected for a decode head the decode head [calls](https://github.com/open-mmlab/mmsegmentation/blob/d35fbbdb47383906399df3a23e4f9f1db1e68b2b/mmseg/models/decode_heads/decode_head.py#L232) the accuracy function [here](https://github.com/open-mmlab/mmsegmentation/blob/d35fbbdb47383906399df3a23e4f9f1db1e68b2b/mmseg/models/losses/accuracy.py#L5). This appears to be the only place that will throw an error with a `gt_semantic_seg` as a 3D tensor of non-mutually exclusive seg maps. 

It would be nice if the requirement on `gt_semantic_seg` being a single mutually exclusive seg map was relaxed or removed to support additional use cases. 

If this is something the maintainers are interested in, we'd be open to creating a PR.

**Comment by xvjiarui** [comment]
Really nice propose. 
We may breakdown it into several PRs. 
Btw, I don't think most of the preprocessing operator like crop support images with channel > 3. How should we deal with that?

**Comment by zachcoleman** [comment]
@MengzhangLI 
So, me and some of my colleagues are open to helping/generating PRs to incorporate a change like this. We did not want to do that work without getting a feel for how interested you are in incorporating something like this in the project first. 


@xvjiarui 

> Btw, I don't think most of the preprocessing operator like crop support images with channel > 3. How should we deal with that?

I don't know what's best for this.

As some background, in our use case we actually use the `mmdet` CocoDataset and the MMDetection dataset pipelines. Then generate the necessary `gt_semantic_seg` to work with `mmseg` models using a custom registered pipeline module. The custom pipeline module iterates over the `gt_masks` and creates a single 3D tensor for `gt_semantic_seg`. 

The easiest option is to probably have cropping and other modules dynamically check the dimension of the `gt_semantic_seg` mask before doing anything. In our custom DefaultFormatBundle we do a check like this as a quick fix:
```python
if "gt_semantic_seg" in results:
    if len(results["gt_semantic_seg"].shape) == 2:
        results["gt_semantic_seg"] = DC(
            to_tensor(results["gt_semantic_seg"][None, ...]), stack=True
        )
    elif len(results["gt_semantic_seg"].shape) == 3:
        results["gt_semantic_seg"] = DC(
            to_tensor(results["gt_semantic_seg"]), stack=True
        )
```
